src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala

```diff
@@ -153,14 +156,32 @@ class S3Emitter(
   def attemptEmit(namedStream: NamedStream, connectionAttemptStartTime: Long): Boolean = {
 
     var attemptCount: Long = 1
+    val connectionAttemptStartDateTime = new DateTime(connectionAttemptStartTime)
     while (true) {
       if (attemptCount > 1 && System.currentTimeMillis() - connectionAttemptStartTime > maxConnectionTime) {
         forceShutdown()
       }
 
       try {
         val outputStream = namedStream.stream
-        val filename = namedStream.filename
+        // replace the pattern with actual date values
+        val detectBracesExpression = "\\{(.*?)\\}".r
+        val replacements = detectBracesExpression
+          .findAllMatchIn(directoryPattern.getOrElse(""))
+          .map(_.toString.trim)
+          .map(str => {
+            val converted = Try(DateTimeFormat.forPattern(str).withZone(DateTimeZone.UTC).print(connectionAttemptStartDateTime))
+            (str, converted.getOrElse(str))
+          }
+          )
+          .toList
+        val directory = replacements.foldLeft(directoryPattern.getOrElse("")) {
+          (dir, replacement) => dir.replace(replacement._1, replacement._2)
+        }
+        // ensure trailing slash in the directory only if a pattern was supplied
+        val filename = if (directory.endsWith("/") || directory.isEmpty) s"$directory${namedStream.filename}"
+        else s"$directory/${namedStream.filename}"
```

Comment:
can we extract all this into its own standalone function and add tests?

Reply:
Refactored s3 path calculation to a separate method and added unit tests